### PR TITLE
MTM-61476: Add section describing how to use mqtt service error topic

### PR DESCRIPTION
**Pull request by abor-c8y**

### CODE REVIEW

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 1]
```diff
@@ -59,9 +59,20 @@ MQTT 3.1 and 3.1.1 clients:
 for the SUBSCRIBE packets, where the reason code is `0x80`.
 * For the PUBLISH packets, the client will be disconnected with no further information as per the MQTT specification.
 
-{{< c8y-admon-info >}}
-Work to enable an error topic so that information regarding topic limit failures can be accessible to MQTT 3 clients is underway.
-{{< /c8y-admon-info >}}
+#### Error Topic {#error-topic}
+
+MQTT Service provides clients the ability to review errors through messages received by subscribing to the error topic, `$debug/$error`.
+When subscribing to the topic it will act as a per-client topic in that clients will only receive messages relating to their client id. For example
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md:65]:
```suggestion
When subscribing to the topic it will act as a per-client topic in that clients will only receive messages relating to their client id. For example,
```
```suggestion
When subscribing to the topic it will act as a per-client topic in that clients will only receive messages relating to their client id. For example
```
```suggestion
When subscribing to the topic it will act as a per-client topic in that clients will only receive messages relating to their client id. For example
```
```suggestion
When subscribing to the topic it will act as a per-client topic in that clients will only receive messages relating to their client id. For example
```

Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]:
```suggestion
When subscribing to the topic it will act as a per-client topic in that clients will only receive messages relating to their client ID. For example.
```

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 2]
```diff
@@ -59,9 +59,20 @@ MQTT 3.1 and 3.1.1 clients:
...
+if a client was attempting to subscribe to a new topic, and the creation of the topic would exceed the topic limit, only that client would receive an error.
+
+Due to the behaviour of MQTT 3.1 and 3.1.1 there are some edge cases which may result in client being disconnected before receiving messages from the error topic:
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]: @abor-c8y What do you mean when referring to "edge" cases?

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 3]
```diff
@@ -59,9 +59,20 @@ MQTT 3.1 and 3.1.1 clients:
...
+
+* Server implementation does not authorise a PUBLISH from the client.
+* Bits within fixed portions of a packet are incorrectly set e.g. fixed headers. MQTT specification treats the entire packet as malformed and closes the connection.
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]:
```suggestion
* Bits within fixed portions of a packet are incorrectly set, for example, fixed headers. MQTT specification treats the entire packet as malformed and closes the connection.
```

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 4]
```diff
@@ -59,9 +59,20 @@ MQTT 3.1 and 3.1.1 clients:
...
+
+In such instances MQTT clients must reconnect to be able to receive error messages from error topic via subscription. Error messages received after this reconnection
+are from the previous session which can lead to confusion when attempting corrective actions. Therefore it is highly recommended that to build a microservice using
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]:
```suggestion
are from the previous session which can lead to confusion when attempting corrective actions. Therefore, it is highly recommended that to build a microservice using
```

File: content/device-integration/mqtt-service-bundle/implementation.md (hunk 2 above)
Comment (abor-c8y) [reply content/device-integration/mqtt-service-bundle/implementation.md]: In the bullet points below I outlined the edge cases.

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 5]
```diff
@@ -59,9 +59,20 @@ MQTT 3.1 and 3.1.1 clients:
 for the SUBSCRIBE packets, where the reason code is `0x80`.
 * For the PUBLISH packets, the client will be disconnected with no further information as per the MQTT specification.
 
-{{< c8y-admon-info >}}
-Work to enable an error topic so that information regarding topic limit failures can be accessible to MQTT 3 clients is underway.
-{{< /c8y-admon-info >}}
+#### Error Topic {#error-topic}
+
+MQTT Service provides clients the ability to review errors through messages received by subscribing to the error topic, `$debug/$error`.
+When subscribing to the topic it will act as a per-client topic in that clients will only receive messages relating to their client ID. For example
```
Comment (schmiel-cumulocity) [inline content/device-integration/mqtt-service-bundle/implementation.md]: `in that clients` is somehow misleading for me and thus hard to understand, maybe we could rephrase this to something like:
```suggestion
When subscribing to the topic, it will act as a per-client topic, meaning the client will receive messages exclusively related to his client ID only. For example
```

File: content/device-integration/mqtt-service-bundle/implementation.md (hunk 2 above)
Comment (MWindel) [reply content/device-integration/mqtt-service-bundle/implementation.md]: @abor-c8y My question referred to the point if you are referring here to Cumulocity IoT Edge or just cases that are not appearing as often? In the latter case it would be better to use another adjective.

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 6]
```diff
@@ -59,9 +59,20 @@ MQTT 3.1 and 3.1.1 clients:
...
+MQTT Service provides clients the ability to review errors through messages received by subscribing to the error topic, `$debug/$error`.
+When subscribing to the topic it will act as a per-client topic, meaning the client will only receive messages exclusively related to their client ID. For example
+if a client was attempting to subscribe to a new topic, and the creation of the topic would exceed the topic limit, only that client would receive an error.
+
+Due to the behaviour of MQTT 3.1 and 3.1.1 there are some edge cases which may result in client being disconnected before receiving messages from the error topic:
+
+* Server implementation does not authorise a PUBLISH from the client.
+* Bits within fixed portions of a packet are incorrectly set, for example, fixed headers. MQTT specification treats the entire packet as malformed and closes the connection.
+
+In such instances MQTT clients must reconnect to be able to receive error messages from error topic via subscription. Error messages received after this reconnection
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]:
```suggestion
In such instances MQTT clients must reconnect to be able to receive error messages from the error topic via the subscription. Error messages received after this reconnection
```

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 7]
```diff
@@ -59,9 +59,20 @@ MQTT 3.1 and 3.1.1 clients:
...
+are from the previous session which can lead to confusion when attempting corrective actions. Therefore, it is highly recommended to build a microservice which uses
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]:
```suggestion
are from the previous session which can lead to confusion when attempting corrective actions. Therefore, we highly recommend to build a microservice which uses
```

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 8]
```diff
@@ -59,9 +59,20 @@ MQTT 3.1 and 3.1.1 clients:
...
+the MQTT Service SDK to consume error messages, or use MQTT 5 for clients and make use of the reason codes feature.
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md:73]:
```suggestion
the MQTT Service SDK to consume error messages, or use MQTT 5 for clients and make to use of the reason codes feature.
```

File: content/device-integration/mqtt-service-bundle/implementation.md (hunk 2 above)
Comment (schmiel-cumulocity) [reply content/device-integration/mqtt-service-bundle/implementation.md]: Maybe we can make it more generic and skip pointing those corner cases? According to the MQTT Spec
```
4.8 Handling errors
 

Unless stated otherwise, if either the Server or Client encounters a protocol violation, it MUST close the Network Connection on which it received that Control Packet which caused the protocol violation [MQTT-4.8.0-1].

A Client or Server implementation might encounter a Transient Error (for example an internal buffer full condition) that prevents successful processing of an MQTT packet.

If the Client or Server encounters a Transient Error while processing an inbound Control Packet it MUST close the Network Connection on which it received that Control Packet [MQTT-4.8.0-2]. If a Server detects a Transient Error it SHOULD NOT disconnect or have any other effect on its interactions with any other Client.
```
So we could just write something like:
`According to the MQTT 3.1.1 specification, if either the Server or Client encounters a protocol violation, it MUST close the Network Connection on which it received that Control Packet which caused the protocol violation`

**Review by schmiel-cumulocity — APPROVED** [review]

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 9]
```diff
@@ -59,9 +59,18 @@ MQTT 3.1 and 3.1.1 clients:
...
-{{< c8y-admon-info >}}
-Work to enable an error topic so that information regarding topic limit failures can be accessible to MQTT 3 clients is underway.
-{{< /c8y-admon-info >}}
+#### Error Topic {#error-topic}
+
+MQTT Service provides clients the ability to review errors through messages received by subscribing to the error topic, `$debug/$error`.
+When subscribing to the topic it will act as a per-client topic, meaning the client will only receive messages exclusively related to their client ID. For example
+if a client was attempting to subscribe to a new topic, and the creation of the topic would exceed the topic limit, only that client would receive an error.
+
+According to the MQTT 3.1.1 specification, if either the Server or the Client encounters a protocol violation, it MUST close the Network connection on
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]:
```suggestion
According to the MQTT 3.1.1 specification, if either the server or the client encounters a protocol violation, it must close the network connection on
```

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 10]
```diff
@@ -59,9 +59,18 @@ MQTT 3.1 and 3.1.1 clients:
 for the SUBSCRIBE packets, where the reason code is `0x80`.
 * For the PUBLISH packets, the client will be disconnected with no further information as per the MQTT specification.
 
-{{< c8y-admon-info >}}
-Work to enable an error topic so that information regarding topic limit failures can be accessible to MQTT 3 clients is underway.
-{{< /c8y-admon-info >}}
+#### Error Topic {#error-topic}
+
+MQTT Service provides clients the ability to review errors through messages received by subscribing to the error topic, `$debug/$error`.
+When subscribing to the topic it will act as a per-client topic, meaning the client will only receive messages exclusively related to their client ID. For example
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]:
```suggestion
When subscribing to the topic it will act as a per-client topic, meaning the client will only receive messages exclusively related to their client ID. For example,
```

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 11]
```diff
@@ -59,9 +59,18 @@ MQTT 3.1 and 3.1.1 clients:
...
+if a client was attempting to subscribe to a new topic, and the creation of the topic would exceed the topic limit, only that client would receive an error.
+
+According to the MQTT 3.1.1 specification, if either the Server or the Client encounters a protocol violation, it MUST close the Network connection on
+which it received the Control Packet which caused the violation.
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]:
```suggestion
which it received the control packet which caused the violation.
```

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 12]
```diff
@@ -59,9 +59,18 @@ MQTT 3.1 and 3.1.1 clients:
...
+
+In such instances MQTT clients must reconnect to be able to receive error messages from the error topic via the subscription. Error messages received after this reconnection
+are from the previous session which can lead to confusion when attempting corrective actions. Therefore, we highly recommend to build a microservice which uses
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]:
```suggestion
are from the previous session. This can lead to confusion when attempting corrective actions. Therefore, we highly recommend you to build a microservice which uses
```

File: content/device-integration/mqtt-service-bundle/implementation.md [hunk 13]
```diff
@@ -59,9 +59,18 @@ MQTT 3.1 and 3.1.1 clients:
...
+the MQTT Service SDK to consume error messages, or use MQTT 5 for clients and make to use of the reason codes feature.
```
Comment (MWindel) [inline content/device-integration/mqtt-service-bundle/implementation.md]: @abor-c8y This sentence does not make any sense. Please rephrase.

Comment (MWindel) [reply content/device-integration/mqtt-service-bundle/implementation.md]:
```suggestion
the MQTT Service SDK to consume error messages, or use MQTT 5 for clients and make use of the reason codes feature.
```

**Review by MWindel — APPROVED** [review]
